lib/app_profiler.rb

```diff
@@ -60,17 +68,78 @@ module Viewer
   mattr_reader :after_process_queue, default: nil
 
   class << self
-    def run(*args, &block)
-      Profiler.run(*args, &block)
+    def run(*args, backend: nil, **kwargs, &block)
+      orig_backend = self.backend
+      begin
+        self.backend = backend if backend
+        profiler.run(*args, **kwargs, &block)
+      rescue BackendError => e
+        logger.error(
+          "[AppProfiler.run] exception #{e} configuring backend #{backend}: #{e.message}"
+        )
+        yield
+      end
+    ensure
+      AppProfiler.backend = orig_backend
     end
 
     def start(*args)
-      Profiler.start(*args)
+      profiler.start(*args)
     end
 
     def stop
-      Profiler.stop
-      Profiler.results
+      profiler.stop
+      profiler.results
+    end
+
+    def running?
+      @backend&.running?
+    end
+
+    def profiler
+      backend
+      @backend ||= @profiler_backend.new
+    end
+
+    def backend=(new_backend)
+      return if new_backend == backend
+
+      new_profiler_backend = backend_for(new_backend)
+
+      if running?
+        raise BackendError,
+          "cannot change backend to #{new_backend} while #{backend} backend is running"
+      end
+
+      return if @profiler_backend == new_profiler_backend
+
+      clear
+      @profiler_backend = new_profiler_backend
+    end
+
+    def backend_for(backend_name)
+      if vernier_supported? &&
+          backend_name == AppProfiler::Backend::VernierBackend::NAME
+        AppProfiler::Backend::VernierBackend
+      elsif backend_name == AppProfiler::Backend::StackprofBackend::NAME
+        AppProfiler::Backend::StackprofBackend
+      else
+        raise BackendError, "unknown backend #{backend_name}"
+      end
+    end
+
+    def backend
+      @profiler_backend ||= Backend::StackprofBackend
+      @profiler_backend::NAME
+    end
+
+    def vernier_supported?
+      defined?(AppProfiler::Backend::VernierBackend::NAME)
+    end
+
+    def clear
+      @backend.stop if @backend&.running?
```

Comment:
I think this might be overly defensive. If we're trying to assign a new backend and the backend is running, shouldn't we raise instead of stopping?

Reply:
Err wait, i remember why i did this now - it's to prevent a running backend from becoming orphaned via the clear method. If we lose a reference to the backend, and a session is running, we can NEVER end the session. This is to ensure that doesn't happen.

Reply:
Made it private, and added a test.